examples/bridge-app/esp32/main/DeviceCallbacks.cpp

```diff
@@ -0,0 +1,125 @@
+        break;
+
+    case DeviceEventType::kInterfaceIpAddressChanged:
+        if ((event->InterfaceIpAddressChanged.Type == InterfaceIpChangeType::kIpV4_Assigned) ||
+            (event->InterfaceIpAddressChanged.Type == InterfaceIpChangeType::kIpV6_Assigned))
+        {
+            // MDNS server restart on any ip assignment: if link local ipv6 is configured, that
+            // will not trigger a 'internet connectivity change' as there is no internet
+            // connectivity. MDNS still wants to refresh its listening interfaces to include the
+            // newly selected address.
+            chip::app::Mdns::StartServer();
+        }
+        break;
+    }
+
+    ESP_LOGI(TAG, "Current free heap: %d\n", heap_caps_get_free_size(MALLOC_CAP_8BIT));
+}
+void DeviceCallbacks::PostAttributeChangeCallback(EndpointId endpointId, ClusterId clusterId, AttributeId attributeId, uint8_t mask,
+                                                  uint16_t manufacturerCode, uint8_t type, uint16_t size, uint8_t * value)
+{
+    ESP_LOGI(TAG, "PostAttributeChangeCallback - Cluster ID: '0x%04x', EndPoint ID: '0x%02x', Attribute ID: '0x%04x'", clusterId,
+             endpointId, attributeId);
+
+    switch (clusterId)
+    {
+    case ZCL_ON_OFF_CLUSTER_ID:
+        OnOnOffPostAttributeChangeCallback(endpointId, attributeId, value);
+        break;
+
+    default:
+        ESP_LOGI(TAG, "Unhandled cluster ID: %d", clusterId);
+        break;
+    }
+
+    ESP_LOGI(TAG, "Current free heap: %d\n", heap_caps_get_free_size(MALLOC_CAP_8BIT));
+}
+
+void DeviceCallbacks::OnInternetConnectivityChange(const ChipDeviceEvent * event)
+{
+    if (event->InternetConnectivityChange.IPv4 == kConnectivity_Established)
+    {
+        ESP_LOGI(TAG, "Server ready at: %s:%d", event->InternetConnectivityChange.address, CHIP_PORT);
+        chip::app::Mdns::StartServer();
+    }
+    else if (event->InternetConnectivityChange.IPv4 == kConnectivity_Lost)
+    {
+        ESP_LOGE(TAG, "Lost IPv4 connectivity...");
+    }
+    if (event->InternetConnectivityChange.IPv6 == kConnectivity_Established)
+    {
+        ESP_LOGI(TAG, "IPv6 Server ready...");
+        chip::app::Mdns::StartServer();
+    }
+    else if (event->InternetConnectivityChange.IPv6 == kConnectivity_Lost)
+    {
+        ESP_LOGE(TAG, "Lost IPv6 connectivity...");
+    }
+}
+
+void DeviceCallbacks::OnSessionEstablished(const ChipDeviceEvent * event)
+{
+    if (event->SessionEstablished.IsCommissioner)
+    {
+        ESP_LOGI(TAG, "Commissioner detected!");
+    }
+}
+
+void DeviceCallbacks::OnOnOffPostAttributeChangeCallback(EndpointId endpointId, AttributeId attributeId, uint8_t * value)
+{
+    VerifyOrExit(attributeId == ZCL_ON_OFF_ATTRIBUTE_ID, ESP_LOGI(TAG, "Unhandled Attribute ID: '0x%04x", attributeId));
+    VerifyOrExit(endpointId == 1 || endpointId == 2, ESP_LOGE(TAG, "Unexpected EndPoint ID: `0x%02x'", endpointId));
+    /*Actions on fixed endpoints*/
+exit:
```

Comment:
please use return macros instead of exit macros, since this exit does not contain any cleanup code.

it may be more readable to just use if considitons and no macros at all:

```cpp

if (attributeId != ZCL_ON_OFF_ATTRIBUTE_ID) {
  // is logI correct here, or should it be an error?
  ESP_LOGI(TAG, "Unhandled Attribute ID: '0x%04x", attributeId);
  return;
}

if ((endpointId < 1) || (endpointId > 2)) {
  ESP_LOGE(TAG, "Unexpected EndPoint ID: `0x%02x'", endpointId);
  return;
}

Reply:
Done